### PR TITLE
[depends] patch spdlog to require at least fmt 5.3.0

### PR DESCRIPTION
**Pull request by Montellese**
## Description
`spdlog` links against the `fmt::fmt` CMake target which was introduced with `fmt 5.3.0`.

## Motivation and Context
Right now Kodi might pick up `fmt` versions installed on the system (e.g. through `apt-get`)  which are older `3.X.Y` on Ubuntu 16.04 or `4.0.0` on Ubuntu 18.04 which are not compatible at all.

I'm trying to get this upstreamed, see https://github.com/gabime/spdlog/pull/1525.

## How Has This Been Tested?
* I have manually built on Ubuntu 18.04 with and without `libfmt-dev 4.0.0` on my system
* @MilhouseVH has tested this on Ubuntu 19.10 with `libfmt3-dev` installed on his system

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: tools/depends/target/libspdlog/0001-fix_fmt_version.patch [hunk 1]
```diff
@@ -0,0 +1,11 @@
+--- ae/CMakeLists.txt	2020-01-13 09:06:56.000000000 +0100
++++ b/CMakeLists.txt	2020-04-20 13:29:47.569870108 +0200
+@@ -147,7 +147,7 @@ target_link_libraries(spdlog_header_only
```
Comment (Rechi) [inline tools/depends/target/libspdlog/0001-fix_fmt_version.patch:3]:
```suggestion
--- ae/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -147,7 +147,7 @@
```

**Comment by asavah** [comment]
AFAIK it should be at least higher than 5.1.0 see https://forum.kodi.tv/showthread.php?tid=353599

**Review by Rechi — COMMENTED** [review]
Please follow the patch style used everywhere else.

**Review by wsnipex — APPROVED** [review]
looks sane

File: tools/depends/target/libspdlog/0001-fix_fmt_version.patch (hunk 1 above)
Comment (Montellese) [reply tools/depends/target/libspdlog/0001-fix_fmt_version.patch]: Sorry for the typo. Should be fixed.

**Comment by Rechi** [comment]
There is another issue. If a valid fmt version is installed and both internal fmt and spdlog are enabled, Kodi uses the internal fmt version, but spdlog uses the sytem fmt version.

**Comment by Montellese** [comment]
I changed the patch to specify version `5.3.0` for `fmt` based on my observations listed in https://github.com/gabime/spdlog/pull/1525#issuecomment-616717542.

**Comment by Montellese** [comment]
> There is another issue. If a valid fmt version is installed and both internal fmt and spdlog are enabled, Kodi uses the internal fmt version, but spdlog uses the sytem fmt version.

I've added a second commit which should address this issue but I have no idea if it is the right way to do it. I'm passing `CMAKE_PREFIX_PATH` to the `externalproject_add(spdlog)` call with `CMAKE_PREFIX_PATH` pointing at the internally built `fmt` if `-DENABLE_INTERNAL_FMT=ON` has been specified.

I tried setting `fmt_ROOT` instead which would be way more specific than `CMAKE_PREFIX_PATH` but that didn't work and it seems that `fmt_ROOT` is only valid as an environment variable.

With this I can have `libfmt 4.0.0` on my system and the patch from the first commit reverted and it still finds the proper (internally built) `fmt`.

File: cmake/modules/FindSpdlog.cmake [hunk 2]
```diff
@@ -35,6 +35,12 @@ if(ENABLE_INTERNAL_SPDLOG)
 
   set(SPDLOG_LIBRARY ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/libspdlog.a)
   set(SPDLOG_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/include)
+
+  # make sure we use the internally built fmt
+  if(ENABLE_INTERNAL_FMT)
+    set(FMT_ROOT ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR})
```
Comment (Rechi) [inline cmake/modules/FindSpdlog.cmake:41]: Why only set this if internal fmt is enabled? Always setting it doesn't cause any harm.

**Comment by Rechi** [comment]
The same issue also happens for the 'native' building of `tools/depends/target/lib{fmt,spdlog}` if a non standard PREFIX is used.

**Comment by Montellese** [comment]
> The same issue also happens for the 'native' building of `tools/depends/target/lib{fmt,spdlog}` if a non standard PREFIX is used.

How should we know what prefix wad used to build `libfmt`? Theoretically you can use different prefixes for each native build. Should we just inject the prefix used for building `libspdlog`?

File: cmake/modules/FindSpdlog.cmake (hunk 2 above)
Comment (Montellese) [reply cmake/modules/FindSpdlog.cmake:41]: I'm not sure of the side effects. But any library found in the internal build directory should probably have precedence over system libraries.

**Comment by Rechi** [comment]
I was only thinking about building libfmt and libsplog (in the correct order) into the same prefix.
If you are using different prefixes you actually want that the libs are independent. Then they should not be considered, unless libfmt is installed into a system prefix, but cmake does that by default anyway.

File: cmake/modules/FindSpdlog.cmake (hunk 2 above)
Comment (Montellese) [reply cmake/modules/FindSpdlog.cmake]: I removed the `if(ENABLE_INTERNAL_FMT)`

**Comment by Montellese** [comment]
> I was only thinking about building libfmt and libsplog (in the correct order) into the same prefix.
> If you are using different prefixes you actually want that the libs are independent. Then they should not be considered, unless libfmt is installed into a system prefix, but cmake does that by default anyway.

I've added the `-DCMAKE_PREFIX_PATH=${PREFIX}` to `tools/depends/target/libspdlog/Makefile`.

**Comment by Montellese** [comment]
BTW specifying the `5.3.0` version for `find_package(fmt 5.3.0 REQUIRED)` has been accepted upstream.

**Comment by basilgello** [comment]
LGTM. Is there anything preventing this from being merged atm?

File: tools/depends/target/libspdlog/Makefile [hunk 3]
```diff
@@ -12,7 +12,8 @@ CMAKE_OPTIONS= \
 	-DSPDLOG_BUILD_TESTS=OFF \
 	-DSPDLOG_BUILD_BENCH=OFF \
 	-DSPDLOG_INSTALL=ON \
-	-DSPDLOG_FMT_EXTERNAL=ON
+	-DSPDLOG_FMT_EXTERNAL=ON \
+	-DCMAKE_PREFIX_PATH=${PREFIX}
```
Comment (Rechi) [inline tools/depends/target/libspdlog/Makefile:16]: I'm not sure if it's a good idea to pass this for full depends builds too. I would add it to the CMAKE variable in line 31.

Comment (Montellese) [reply tools/depends/target/libspdlog/Makefile:16]: You're right I actually wanted to do that but must have ended up at the wrong place.

Comment (Montellese) [reply tools/depends/target/libspdlog/Makefile]: Fixed

**Comment by Montellese** [comment]
> LGTM. Is there anything preventing this from being merged atm?

I'm waiting on final feedback from @Rechi and @lrusak which know more about this whole depends system as me.

File: tools/depends/target/libspdlog/Makefile [hunk 4]
```diff
@@ -28,7 +28,7 @@ else
     RETRIEVE_TOOL_FLAGS := -Ls --create-dirs -f -O
     ARCHIVE_TOOL := tar
     ARCHIVE_TOOL_FLAGS := --strip-components=1 -xf
-    CMAKE := cmake -DCMAKE_INSTALL_PREFIX=$(PREFIX)
+    CMAKE := cmake -DCMAKE_INSTALL_PREFIX=$(PREFIX) -DCMAKE_PREFIX_PATH=${PREFIX}
```
Comment (Rechi) [inline tools/depends/target/libspdlog/Makefile:31]: Does `${PREFIX}` work the same as `$(PREFIX)` and is there any reason for not using the same style for both?

Comment (wsnipex) [reply tools/depends/target/libspdlog/Makefile:31]: yes, it's the same. We should still try to use the same style

Comment (phunkyfish) [reply tools/depends/target/libspdlog/Makefile:31]: So the preferred style is `$(PREFIX)`? So using round brackets. Good to know.

Comment (Montellese) [reply tools/depends/target/libspdlog/Makefile]: Sorry too much `cmake` syntax recently ;-) Fixed.

**Comment by popcornmix** [comment]
I can also confirm that kodi master doesn't build on debian buster, but with this PR it does.

**Comment by basilgello** [comment]
@popcornmix Do you build Debian packages for buster? I am preparing the Debian source packages [here](https://salsa.debian.org/basilgello-guest) and it would be great to have some more testers for it. I do use Kodi + inputstreams + iptvsimple only, and there are ~120 packages in total.

**Comment by Montellese** [comment]
jenkins build this please

**Comment by fuzzard** [comment]
Jenkins build this please

**Comment by basilgello** [comment]
Team merge this please

:laughing:

**Pull request merged**